File: c/src/snake.c
#include "snake.h"

#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>

bool position_equals(Position a, Position b) {
    return a.x == b.x && a.y == b.y;
}

Snake *make_snake(Position start) {
    Snake *s = malloc(sizeof(Snake));
    if (s == NULL) {
        return NULL;
    }

    s->length = 1;
    s->body[0] = start;
    s->direction = RIGHT;

    return s;
}

void move_snake(Snake *s, Direction dir) {
    for (int i = s->length - 1; i > 0; i--) {
        s->body[i] = s->body[i - 1];
    }

    switch (dir) {
        case UP:
            s->body[0].y--;
            break;
        case DOWN:
            s->body[0].y++;
            break;
        case LEFT:
            s->body[0].x--;
            break;
        case RIGHT:
            s->body[0].x++;
            break;
    }

    s->direction = dir;
}

bool snake_contains_position(Snake *s, Position pos) {
    for (int i = 0; i < s->length; i++) {
        if (position_equals(s->body[i], pos)) {
            return true;
        }
    }

    return false;
}

bool snake_collided_with_self(Snake *s) {
    Position head = s->body[0];
    for (int i = 1; i < s->length; i++) {
        if (position_equals(s->body[i], head)) {
            return true;
        }
    }

    return false;
}

bool snake_collided_with_wall(Snake *s, int max_x, int max_y) {
    Position head = s->body[0];
    return head.x <= 0 || head.x >= max_x - 1 || head.y <= 0 ||
           head.y >= max_y - 1;
}

File: rust/src/snake.rs
/// A simple position struct to track coordinates on the terminal.
#[derive(Clone, Copy, PartialEq)]
pub struct Position {
    pub x: u16,
    pub y: u16,
}

/// Possible movement directions.
#[derive(Clone, Copy)]
pub enum Direction {
    Up,
    Down,
    Left,
    Right,
}

/// The Snake is represented by a vector of Positions (its body) and a current direction.
pub struct Snake {
    pub body: Vec<Position>,
    direction: Direction,
}

impl Snake {
    pub(crate) fn new(start: Position) -> Self {
        Self {
            body: vec![start],
            direction: Direction::Right,
        }
    }

    /// Returns the current head (first element) of the snake.
    pub(crate) fn head(&self) -> Position {
        self.body[0]
    }

    /// Moves the snake one step forward in its current direction.
    /// If `grow` is true, the snake's tail is not removed (thus growing the snake).
    pub(crate) fn move_forward(&mut self, grow: bool) {
        let head = self.head();
        let new_head = match self.direction {
            Direction::Up => Position {
                x: head.x,
                y: head.y.saturating_sub(1),
            },
            Direction::Down => Position {
                x: head.x,
                y: head.y + 1,
            },
            Direction::Left => Position {
                x: head.x.saturating_sub(1),
                y: head.y,
            },
            Direction::Right => Position {
                x: head.x + 1,
                y: head.y,
            },
        };

        self.body.insert(0, new_head);
        if !grow {
            self.body.pop();
        }
    }

    /// Change the snake's direction. Prevents 180° reversal.
    pub(crate) fn change_direction(&mut self, new_direction: Direction) {
        // Prevent reverse direction (e.g., cannot go from Up to Down directly).
        match (&self.direction, &new_direction) {
            (Direction::Up, Direction::Down)
            | (Direction::Down, Direction::Up)
            | (Direction::Left, Direction::Right)
            | (Direction::Right, Direction::Left) => {}
            _ => self.direction = new_direction,
        }
    }

    /// Checks if a given position is occupied by the snake.
    pub(crate) fn is_colliding(&self, pos: &Position) -> bool {
        self.body.iter().any(|&p| p == *pos)
    }

    /// Returns `true` if the snake's head collides with any other part of its body.
    pub(crate) fn check_self_collision(&self) -> bool {
        let head = self.head();
        self.body.iter().skip(1).any(|&p| p == head)
    }
}
